mypy/typeops.py

```diff
@@ -338,9 +340,17 @@ class B(A): pass
 
         # Update the method signature with the solutions found.
         # Technically, some constraints might be unsolvable, make them Never.
-        to_apply = [t if t is not None else UninhabitedType() for t in typeargs]
-        func = expand_type(func, {tv.id: arg for tv, arg in zip(self_vars, to_apply)})
-        variables = [v for v in func.variables if v not in self_vars]
+        # (... unless this is `__init__`, because we're going to be returning
+        #  the self-type anyways.)
+        to_apply = [
+            (tv, t) if t is not None else (tv, UninhabitedType())
+            for tv, t in zip(self_vars, typeargs)
+            if not is_init
```

Comment:
Should this special casing only apply to type vars in `self_ids` or something? (If so, would be good to have test case that catches this)

Reply:
Well all the vars in `self_vars` are also in `self_ids`.

Ideally my first attempt should work + just binding `Self` but I'm not exactly sure where to find `Self`.

Reply:
Oh oops, I just read code sloppily, thought we had other things from `func.variables`